cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -143,39 +142,202 @@ TEST(ScalarTemporalTest, TestTemporalComponentExtractionWithDifferentUnits) {
+  CheckScalarUnary("day_of_week", unit, times, int64(), day_of_week);
+  CheckScalarUnary("day_of_year", unit, times, int64(), day_of_year);
+  CheckScalarUnary("iso_year", unit, times, int64(), iso_year);
+  CheckScalarUnary("iso_week", unit, times, int64(), iso_week);
+  CheckScalarUnary("iso_calendar", ArrayFromJSON(unit, times), iso_calendar);
+  CheckScalarUnary("quarter", unit, times, int64(), quarter);
+  CheckScalarUnary("hour", unit, times, int64(), hour);
+  CheckScalarUnary("minute", unit, times, int64(), minute);
+  CheckScalarUnary("second", unit, times, int64(), second);
+  CheckScalarUnary("millisecond", unit, times, int64(), millisecond);
+  CheckScalarUnary("microsecond", unit, times, int64(), microsecond);
+  CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
+  CheckScalarUnary("subsecond", unit, times, float64(), subsecond);
+}
+
+TEST(ScalarTemporalTest, TestOverflow) {
+  const char* times = R"(["1677-09-20T00:00:59.123456789", "2262-04-13T23:23:23.999999999"])";
+
+  auto unit = timestamp(TimeUnit::NANO);
+  auto iso_calendar_type =
+      struct_({field("iso_year", int64()), field("iso_week", int64()),
+               field("iso_day_of_week", int64())});
+
+  auto year = "[2262, 1677]";
+  auto month = "[4, 9]";
+  auto day = "[10, 22]";
+  auto day_of_week = "[3, 2]";
+  auto day_of_year = "[100, 265]";
+  auto iso_year = "[2262, 1677]";
+  auto iso_week = "[15, 38]";
+  auto iso_calendar =
+      ArrayFromJSON(iso_calendar_type,
+                    R"([{"iso_year": 2262, "iso_week": 15, "iso_day_of_week": 4},
+                        {"iso_year": 1677, "iso_week": 38, "iso_day_of_week": 3}])");
+  auto quarter = "[2, 3]";
+  auto hour = "[23, 23]";
+  auto minute = "[35, 48]";
+  auto second = "[32, 50]";
+  auto millisecond = "[833, 290]";
+  auto microsecond = "[8, 448]";
+  auto nanosecond = "[405, 383]";
+  auto subsecond = "[0.833008405, 0.290448383]";
+
+  CheckScalarUnary("year", unit, times, int64(), year);
+  CheckScalarUnary("month", unit, times, int64(), month);
+  CheckScalarUnary("day", unit, times, int64(), day);
+  CheckScalarUnary("day_of_week", unit, times, int64(), day_of_week);
+  CheckScalarUnary("day_of_year", unit, times, int64(), day_of_year);
+  CheckScalarUnary("iso_year", unit, times, int64(), iso_year);
+  CheckScalarUnary("iso_week", unit, times, int64(), iso_week);
+  CheckScalarUnary("iso_calendar", ArrayFromJSON(unit, times), iso_calendar);
+  CheckScalarUnary("quarter", unit, times, int64(), quarter);
+  CheckScalarUnary("hour", unit, times, int64(), hour);
+  CheckScalarUnary("minute", unit, times, int64(), minute);
+  CheckScalarUnary("second", unit, times, int64(), second);
+  CheckScalarUnary("millisecond", unit, times, int64(), millisecond);
+  CheckScalarUnary("microsecond", unit, times, int64(), microsecond);
+  CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
+  CheckScalarUnary("subsecond", unit, times, float64(), subsecond);
 }
```

Comment:
It this what you had in mind [here](https://github.com/apache/arrow/pull/10176#issuecomment-858392898).

Reply:
Yes, thanks!

Reply:
Apart from that you should be using a different unit as NANO?

Reply:
Oh, now I understand the intent.

Reply:
Done.
Perhaps `TestOverflow` id not the best name for the test?